### PR TITLE
[SPARK-39068][SQL] Make thriftserver and sparksql-cli support in-memory catalog

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

if users specify in-memory explicitly, we bypass enable hive support.

### Why are the changes needed?

- with in-memory catalog, we can start multiple instances in the same folder without the limitation[1] of derby.
- spark now is multi catalog architecture, the built-in catalog is not always necessary, in-memory is much lightweight if it is unused. if we use hive, in order to pass [1], we need a extra metastore service

### Does this PR introduce _any_ user-facing change?

yes, `spark.sql.catalogImplementation=in-memory` can be used for ./bin/spark-sql; and thriftserver to enable a lightweight in-memory catalog

### How was this patch tested?

add a new test

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @HyukjinKwon @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
cc @wangyum FYI

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala [hunk 1]
```diff
@@ -48,11 +50,18 @@ private[hive] object SparkSQLEnv extends Logging {
         .setAppName(maybeAppName.getOrElse(s"SparkSQL::${Utils.localHostName()}"))
         .set(SQLConf.DATETIME_JAVA8API_ENABLED, true)
 
+      // if user specified in-memory explicitly, we bypass enable hive support.
+      val shouldUseInmemoryCatalog =
```
Comment (wangyum) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala:54]: `shouldUseInmemoryCatalog` -> `shouldUseInMemoryCatalog`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 2]
```diff
@@ -680,4 +681,22 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
     sessionState.close()
     SparkSQLEnv.stop()
   }
+
+  test("SPARK-39068: support in-memory catalog and running concurrently") {
+    val extraConf = Seq("-c", "spark.sql.catalogImplementation=in-memory")
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:686]: `spark.sql.catalogImplementation=in-memory` -> `${StaticSQLConf.CATALOG_IMPLEMENTATION.key}=in-memory`?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: thanks updated

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala]: thanks, updated

**Review by wangyum — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**